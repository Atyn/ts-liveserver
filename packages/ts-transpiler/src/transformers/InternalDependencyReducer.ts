import TypeScript from 'typescript'

const INTERNAL_PACKAGES = ['path', 'util']

/*
Removed require('path')
*/
export default class InternalDependencyReducer
	implements TypeScript.CustomTransformer
{
	private context: TypeScript.TransformationContext
	constructor(context: TypeScript.TransformationContext) {
		this.context = context
	}
	transformSourceFile(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] => {
			if (
				TypeScript.isTryStatement(node) &&
				node.tryBlock &&
				this.shouldCrash(node.tryBlock)
			) {
				if (
					node?.catchClause?.block &&
					TypeScript.isBlock(node.catchClause.block) &&
					node.catchClause.block.statements.length
				) {
					return [...node.catchClause.block.statements]
				}
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	transformBundle(node: TypeScript.Bundle): TypeScript.Bundle {
		return node
	}
	private shouldCrash(block: TypeScript.Block): boolean {
		let crashing = false
		const visit = (node: TypeScript.Node): TypeScript.Node => {
			if (
				TypeScript.isCallExpression(node) &&
				node.expression &&
				TypeScript.isIdentifier(node.expression) &&
				node.expression.text === 'require' &&
				node.arguments
			) {
				const argument = node.arguments[0]
				if (
					TypeScript.isStringLiteral(argument) &&
					argument &&
					INTERNAL_PACKAGES.includes(argument.text)
				) {
					crashing = true
					return node
				}
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		TypeScript.visitEachChild(block, visit, this.context)
		return crashing
	}
}
