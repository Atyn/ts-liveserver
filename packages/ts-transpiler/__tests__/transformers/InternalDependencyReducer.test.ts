import InternalDependencyReducer from '../../src/transformers/InternalDependencyReducer'
import TypeScript from 'typescript'

describe('InternalDependencyReducer', () => {
	it('Should not have try-catch-statement', async () => {
		const input = `
			try {
				var util = require('util');
				if (typeof util.inherits !== 'function') throw '';
				module.exports = util.inherits;
			} catch (e) {
				module.exports = require('./inherits_browser.js');
			}
		`
		expect(await transformWithPlugin(input)).toMatchSnapshot()
	})
})

const compilerOptions: TypeScript.CompilerOptions = {
	allowJs: true,
	jsx: TypeScript.JsxEmit.React,
	checkJs: false,
	noResolve: true,
	esModuleInterop: true,
	skipLibCheck: true,
	pretty: true,
	allowUnreachableCode: true,
	target: TypeScript.ScriptTarget.ES2020,
	declaration: false,
	module: TypeScript.ModuleKind.ES2020,
	moduleResolution: TypeScript.ModuleResolutionKind.NodeJs,
	sourceMap: false,
}

const transformers: TypeScript.CustomTransformers = {
	after: [(context) => new InternalDependencyReducer(context)],
}

async function transformWithPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
		transformers: transformers,
	})
	return results.outputText.trim()
}
