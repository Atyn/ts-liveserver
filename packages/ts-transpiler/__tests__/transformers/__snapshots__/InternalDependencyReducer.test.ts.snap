// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalDependencyReducer Should not have try-catch-statement 1`] = `"module.exports = require('./inherits_browser.js');"`;
